src/Skylight.Server/Game/Catalog/CatalogOffer.cs

```diff
@@ -44,20 +47,35 @@ internal CatalogOffer(int id, string name, int orderNum, int clubRank, int costC
 		this.HasOffer = hasOffer;
 
 		this.Products = products;
+
+		this.priceCurrency = priceCurrency;
 	}
 
-	public async ValueTask PurchaseAsync(ICatalogTransaction transaction, CancellationToken cancellationToken)
+	private bool CanPurchase(ICatalogTransaction tx) =>
+		tx.GetCurrencyBalance(this.priceCurrency) >= this.CostCredits;
+
+	public async ValueTask PurchaseAsync(
+		ICatalogTransaction transaction,
+		CancellationToken cancellationToken)
 	{
-		foreach (ICatalogProduct product in this.Products)
+		if (this.CostCredits < 0)
 		{
-			ValueTask task = product.PurchaseAsync(transaction, cancellationToken);
+			throw new InvalidOperationException("Offer cost must be non-negative.");
+		}
 
-			if (task.IsCompletedSuccessfully)
-			{
-				continue;
-			}
+		if (!this.CanPurchase(transaction))
+		{
+			throw new InvalidOperationException("Insufficient credits for this purchase.");
```

Comment:
Using exceptions for control flow is discouraged. Validation should be run beforehand so we can short-circuit early.